### PR TITLE
Restore -static-executable on Linux.

### PR DESCRIPTION
**Pull request by spevans**
- Add back StaticBinaryELF.cpp which provides swift::lookupSymbol()
  for statically linked ELF binaries.

- Build libswiftImageInspectionStatic.a from StaticBinaryELF.cpp

- Update static-executable-args.lnk and replace
  libswiftImageInspectionShared with libswiftImageInspectionStatic.

- Add driver tests for '-static-executable' on Linux to validate
  that output binary is statically linked.

- Fix linkage to pthreads to avoid weak linkage issues in a static
  executable.

This is a rebase of a previous PR but that didnt work with Ubutnu18.04 only 16.04. This version has been tested with 16.04, 18.04 and 19.10.

Note, as with `-static-stdlib`, only `Dispatch` can be used with `-static-executable`, `Foundation` cannot at this time.

Integration tests have also been added to catch any future breakage of this change.

### CODE REVIEW

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test

**Comment by spevans** [comment]
cc @jckarter @weissi

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+#define ELFCLASS ELFCLASS32
+typedef Elf32_Ehdr Elf_Ehdr;
+typedef Elf32_Shdr Elf_Shdr;
+typedef Elf32_Phdr Elf_Phdr;
+typedef Elf32_Addr Elf_Addr;
+typedef Elf32_Word Elf_Word;
+typedef Elf32_Sym Elf_Sym;
+typedef Elf32_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF32_ST_TYPE(x)
+#endif
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: I think that using a templated trait type might be better.

```
template <typename Bits_>
struct ELFTraits;

template <>
struct ELFTraits<4> {
  using Header = Elf32_Ehdr;
  using SectionHeader = Elf32_Shdr;
  ...
  static inline constexpr uint32_t symbol_type(value) {
    return ELF32_ST_TYPE(value);
  }
};

template <>
struct ELFTraits<8> {
  ...
};

using ELF = ELFTraits<__SIZEOF_POINTER__>;
```

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Parse a static ELF binary to implement lookupSymbol() address lookup.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__) && defined(__linux__)
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: Why specific to Linux?  Should this not apply to any ELF type?

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_once)(pthread_once_t *, void (*)(void)) = pthread_once;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_key_create)(pthread_key_t *, void (*)(void *)) = pthread_key_create;
+
+
+class StaticBinaryELF {
+
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: Unnecessary whitespace

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+private:
+  // mmap a section of a file that might not be page aligned.
+  class Mapping {
+  public:
+    void *mapping;
+    size_t mapLength;
+    off_t diff;
+
+    Mapping(int fd, size_t fileSize, off_t offset, size_t length) {
+      if (fd < 0 || offset + length > fileSize) {
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: Given that this is internal, perhaps make this a fatal error?  The user should validate that the `fd` and `size` are valid before creating a mapping.

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    return nullptr;
+  }
+
+private:
+  // This is Linux specific - find the full path of the executable
+  // by looking in /proc/self/maps for a mapping holding the current
+  // address space. For a static binary it should only be mapping one
+  // file anyway. Don't use /proc/self/exe as the symlink will be removed
+  // if the main thread terminates - see proc(5).
+  void getExecutablePathName() {
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: Instead of making this entire file linux specific, make this method linux specific and just `#error` on other targets.  This remainder of the code should apply to the BSDs and Haiku.

File: stdlib/public/runtime/ImageInspectionELF.cpp [hunk 6]
```diff
@@ -144,6 +152,8 @@ int swift::lookupSymbol(const void *address, SymbolInfo *info) {
   return 1;
 }
 
+#endif
```
Comment (compnerd) [inline stdlib/public/runtime/ImageInspectionELF.cpp]: Why not just use the alternate version always?  The cost is nearly negligible - the cost is largely associated with the fact that the information from the ELF header parsing *may* be cached by the loader, which is **not** guaranteed.

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 1 above)
Comment (jckarter) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: Maybe we could reuse LLVM's ELFObjectFile headers here?

Comment (compnerd) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: Thats a good idea.

**Comment by spevans** [comment]
@swift-ci test

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 1 above)
Comment (spevans) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: I used the `ELFType` in the end. I had a look at `ELFObjectFile` but it wanted to read in the whole file rather than just the sections that were needed.

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 4 above)
Comment (spevans) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: I made this into an `assert` as that was used elsewhere in the file.

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 5 above)
Comment (spevans) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: Done.

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test linux

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test linux

**Comment by spevans** [comment]
@swift-ci smoke test linux

**Comment by spevans** [comment]
@swift-ci smoke test linux

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test linux

File: lib/Driver/UnixToolChains.cpp [hunk 7]
```diff
@@ -205,6 +200,11 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
     staticStdlib = true;
   }
 
+  if (getTriple().getOS() == llvm::Triple::Linux &&
```
Comment (compnerd) [inline lib/Driver/UnixToolChains.cpp]: Is there a reason for the Linux check?  It should be applicable for any ELFish target

File: stdlib/public/runtime/CMakeLists.txt [hunk 8]
```diff
@@ -102,35 +103,46 @@ if(SWIFT_BUILD_STATIC_STDLIB AND "${sdk}" STREQUAL "LINUX")
...
-      OUTPUT
-        "${LibraryLocation}/${CMAKE_STATIC_LIBRARY_PREFIX}swiftImageInspectionShared${CMAKE_STATIC_LIBRARY_SUFFIX}"
-      DEPENDS
-         ${FragileSupportLibrary})
-    add_dependencies(stdlib ${FragileSupportLibrary})
-    swift_install_in_component(FILES $<TARGET_FILE:${FragileSupportLibrary}>
-                               DESTINATION "lib/swift_static/${lowercase_sdk}/${arch}"
-                               COMPONENT stdlib)
-  endforeach()
+  set(swift_elf_static_lib_compile_flags ${swift_runtime_library_compile_flags})
```
Comment (compnerd) [inline stdlib/public/runtime/CMakeLists.txt]: Please inline this into the single site

File: stdlib/public/runtime/CMakeLists.txt [hunk 9]
```diff
@@ -102,35 +103,46 @@ if(SWIFT_BUILD_STATIC_STDLIB AND "${sdk}" STREQUAL "LINUX")
...
+  list(APPEND swift_elf_static_lib_compile_flags -DELF_STATIC_LIB)
+
+  add_swift_target_library(swiftImageInspectionStatic TARGET_LIBRARY STATIC
+    StaticBinaryELF.cpp ImageInspectionELF.cpp
+    C_COMPILE_FLAGS ${swift_elf_static_lib_compile_flags}
```
Comment (compnerd) [inline stdlib/public/runtime/CMakeLists.txt]: that is: `${swift_runtime_library_compile_flags} -DELF_STATIC_LIB`

File: stdlib/public/runtime/StaticBinaryELF.cpp [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+// as they are normally weak-linked and used to detect the presence of pthreads.
+// Without this the calls just jump to 0x0.
+__attribute__((__visibility__("hidden")))
+ pthread_t (*__strong_pthread_self)(void) = pthread_self;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_once)(pthread_once_t *, void (*)(void)) = pthread_once;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_key_create)(pthread_key_t *, void (*)(void *)) = pthread_key_create;
```
Comment (compnerd) [inline stdlib/public/runtime/StaticBinaryELF.cpp]: I really wish we had linker options ... `-u pthread_self -u pthread_once -u pthead_key_create` is so much better than this.  *sigh*

**Review by compnerd — COMMENTED** [review]
I think I missed it or something ... why not use the parsing logic even for the shared linking?

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 10 above)
Comment (spevans) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: I had used `---defsym=__import_pthread_self=pthread_self` options to the linker before but it looks like that doesnt create strong linkage anymore.

File: lib/Driver/UnixToolChains.cpp (hunk 7 above)
Comment (spevans) [reply lib/Driver/UnixToolChains.cpp]: Lets get it working for one target first. It should be applicable to others but without any testing/ CI its hard to confirm. It should be easier to expand to others later.

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test

File: stdlib/public/runtime/CMakeLists.txt (hunk 8 above)
Comment (spevans) [reply stdlib/public/runtime/CMakeLists.txt]: Fixed.

File: stdlib/public/runtime/CMakeLists.txt (hunk 9 above)
Comment (spevans) [reply stdlib/public/runtime/CMakeLists.txt]: Fixed.

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please smoketest macos

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please smoke test macos

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test macos

**Comment by spevans** [comment]
> I think I missed it or something ... why not use the parsing logic even for the shared linking?

This will only parse 1 ELF file and wont handle the extra shared libraries that are dynamically linked in or their relocations. There would also be no benefit to replacing a platform's `dladdr()` even if a full implementation was written.

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test macos

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 10 above)
Comment (compnerd) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: `--defsym` creating a strong reference sounds like a bug.  If you want to create a strong reference, you should use `--require-defined` or `-u` at the very least.

**Review by compnerd — CHANGES_REQUESTED** [review]
> This will only parse 1 ELF file and wont handle the extra shared libraries that are dynamically linked in or their relocations. There would also be no benefit to replacing a platform's `dladdr()` even if a full implementation was written.

The registration is per module, and is not shared across dynamically shared libraries.  I think that you are focusing too much about the fact that there is a wrapper for the `dladdr` in this file.  That doesn't really belong in there, it was just convenient for it to get placed there I think.

File: stdlib/public/runtime/StaticBinaryELF.cpp (hunk 10 above)
Comment (spevans) [reply stdlib/public/runtime/StaticBinaryELF.cpp]: I didnt know about this option! I had to use `-undefined (-u)` since `ld.gold` is being used, `--require-defined` seems to only work on `ld`

**Comment by spevans** [comment]
Please test with following pull request:
apple/swift-integration-tests#69

@swift-ci please test

**Comment by jckarter** [comment]
`lookupSymbol` should only be used for attempting to symbolicate stack traces, so if it's slow or doesn't work 100% of the time that shouldn't be a showstopper. @drexin had been looking into more robust symbolication using `addr2line` style DWARF line table decoding which might be able to supersede our use of `dladdr` as well.

**Comment by spevans** [comment]
Thats correct, its only really used for stack traces hence why it doesnt matter too much if the `mmap` fails etc. How would the alternative work if there was not DWARF data?

**Comment by spevans** [comment]
@compnerd  FYI: the reason for the  `ImageInspectionShared` / `ImageInspectionStatic` split is mostly due to not wanting to export `dladdr` in the static case. See the discussion in https://github.com/apple/swift/pull/5394 from when the lookupSymbol was added.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by jckarter** [comment]
> Thats correct, its only really used for stack traces hence why it doesnt matter too much if the mmap fails etc. How would the alternative work if there was not DWARF data?

For server side work, I think we'd want to encourage building with at least `-gline-tables-only` as the standard configuration, to enable better quality runtime diagnostics and backtraces. We could still try dladdr as a best effort thing, maybe.

> @compnerd FYI: the reason for the ImageInspectionShared / ImageInspectionStatic split is mostly due to not wanting to export dladdr in the static case. See the discussion in #5394 from when the lookupSymbol was added.

IIRC, at some point `ImageInspectionStatic` also resolved the registration tables completely statically at some point, rather than relying on iterating through images with libdl API that wouldn't be available in a static executable.

**Comment by spevans** [comment]
After checking the man pages and doing a test I found that `dladdr` only finds symbols in `.so` libraries and not in the main executable - this probably explains why the stack traces aren't very good on Linux:

```
$ cat dltest.c
#define _GNU_SOURCE
#include <dlfcn.h>
#include <stdio.h>


void show(void *addr) {
  Dl_info info = {};

  int ret = dladdr(addr, &info);

  if (ret) {
    printf("fname: %s\n", info.dli_fname);
    printf("fbase: %p\n", info.dli_fbase);
    printf("sname: %s\n", info.dli_sname);
    printf("saddr: %p\n\n", info.dli_saddr);
  } else {
    printf("Cant get dladdr\n\n");
  }
}


int main() {
  show(main);
  show(printf);
  return 0;
}

$ clang -Wall -static -O  -o dltest-static dltest.c -ldl
$ ./dltest-static 
Cant get dladdr

Cant get dladdr

$ clang -Wall -O -o dltest-dynamic dltest.c -ldl
$ ./dltest-dynamic 
fname: ./dltest-dynamic
fbase: 0x400000
sname: (null)
saddr: (nil)

fname: ./dltest-dynamic
fbase: 0x400000
sname: printf
saddr: 0x400490
```

So `dladdr` cant be used in a static binary.

**Comment by weissi** [comment]
@spevans I think more than main bin vs library, IIRC `dladdr` can only be used on exported symbols. So it’ll work on printf because that’s exported but it won’t on `main` or other non-exported functions like anything `private` in Swift or a `static` function in C. So I don’t think we can it for anything much in either static or dynamic binaries

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by spevans** [comment]
An updated and more simplified version of this pr is #34180
